module/component/component.go

```diff
@@ -118,76 +117,119 @@ func RunComponent(ctx context.Context, componentFactory ComponentFactory, handle
 			default:
 				panic(fmt.Sprintf("invalid error handling result: %v", result))
 			}
-		case <-done:
-			// Without this additional select, there is a race condition here where the done channel
-			// could have been closed as a result of an irrecoverable error being thrown, so that when
-			// the scheduler yields control back to this goroutine, both channels are available to read
-			// from. If this last case happens to be chosen at random to proceed instead of the one
-			// above, then we would return as if the component shutdown gracefully, when in fact it
-			// encountered an irrecoverable error.
-			select {
-			case err := <-irrecoverableErr:
-				switch result := handler(err); result {
-				case ErrorHandlingRestart:
-					continue
-				case ErrorHandlingStop:
-					return err
-				default:
-					panic(fmt.Sprintf("invalid error handling result: %v", result))
-				}
-			default:
-			}
-
-			// Similarly, the done channel could have closed as a result of the context being canceled.
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			default:
-			}
-
-			// clean completion
-			return nil
+		} else if util.CheckClosed(ctx.Done()) {
+			// the parent context was cancelled
+			stop()
+			return ctx.Err()
 		}
+
+		// clean completion
+		return nil
 	}
 }
 
+// ReadyFunc is called within a ComponentWorker function to indicate that the worker is ready
+// ComponentManager's Ready channel is closed when all workers are ready. This is also the
+// signal used to indicate that the workers returned by lookup are ready.
 type ReadyFunc func()
 
+// LookupFunc returns a ReadyDoneAware interface for the worker identified by id, and a boolean
+// signalling if the worker exists.
+// workerID can be prefixed with ! to get a ReadyDoneAware interface for the collection of all
+// workers EXCEPT the id provided. This is particularly useful if there is a main worker that has
+// shutdown logic that should only run after all other workers have shutdown.
```

Comment:
So, the more appropriate / general solution here is to allow the lookup string to be a regex. That would subsume this usecase, as well as all other usecases.

Again, this temporary solution is just fine for now, but let's create an issue to track this if we want to update it in a future PR.

Reply:
lookups have been removed